public/app/ui/components/search/queryBuilder/QueryBuilderComponent.js

```diff
@@ -0,0 +1,52 @@
+// Based on http://kindohm.github.io/knockout-query-builder/js/viewModel.js
+(function () {
+    'use strict';
+
+    define(
+        [
+            'knockout',
+            'models/query/Group',
+            'models/query/Condition'
+        ],
+        function (ko, Group, Condition) {
+            return function (parameters) {
+                var root;
+
+                if (!parameters.queryObservable || !ko.isObservable(parameters.queryObservable)) {
+                    throw new Error('QueryBuilderComponent missing or invalid required parameter: `queryObservable`.');
+                }
+                if (!parameters.fields || parameters.fields().length === 0) {
+                    throw new Error('QueryBuilderComponent missing required parameter: `fields`.');
+                }
+                if (!parameters.maxDepth) {
+                    throw new Error('QueryBuilderComponent missing required parameter: `maxDepth`.');
+                }
+
+                root = new Group(this);
+                root.parse(parameters.queryObservable());
+                root.query.subscribe(parameters.queryObservable);
```

Comment:
Any changes to the whole tree of rules will change the value of `root.query`.  So this means, whenever anything in the query builder changes, update the query.  This allows changes to move between basic and advanced modes.